pkg/gatewaysecret/handler.go

```diff
@@ -0,0 +1,137 @@
+	gwSecret, err := gsh.findGatewaySecret()
+
+	if isNotFound(err) {
+		return gsh.handleNonExisting(rootSecret)
+	}
+	if err != nil {
+		return err
+	}
+
+	return gsh.handleExisting(rootSecret, gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleNonExisting(rootSecret *apicorev1.Secret) error {
+	gwSecret := gsh.newGatewaySecret(rootSecret)
+	return gsh.create(context.Background(), gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleExisting(rootSecret *apicorev1.Secret, gwSecret *apicorev1.Secret) error {
+	caCert := certmanagerv1.Certificate{}
+	if err := gsh.kcpClient.Get(context.Background(),
+		client.ObjectKey{Namespace: istioNamespace, Name: kcpCACertName},
+		&caCert); err != nil {
+		return fmt.Errorf("failed to get CA certificate: %w", err)
+	}
+
+	if gwSecretLastModifiedAtValue, ok := gwSecret.Annotations[LastModifiedAtAnnotation]; ok {
+		if gwSecretLastModifiedAt, err := time.Parse(time.RFC3339, gwSecretLastModifiedAtValue); err == nil {
+			if caCert.Status.NotBefore != nil && gwSecretLastModifiedAt.After(caCert.Status.NotBefore.Time) {
+				return nil
+			}
+		}
+	}
+
+	gwSecret.Data["tls.crt"] = rootSecret.Data["tls.crt"]
+	gwSecret.Data["tls.key"] = rootSecret.Data["tls.key"]
+	gwSecret.Data["ca.crt"] = rootSecret.Data["ca.crt"]
+	return gsh.update(context.Background(), gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) findGatewaySecret() (*apicorev1.Secret, error) {
+	return gsh.findSecret(context.Background(), client.ObjectKey{
+		Name:      GatewaySecretName,
+		Namespace: istioNamespace,
+	})
+}
+
+func (gsh *GatewaySecretHandler) newGatewaySecret(rootSecret *apicorev1.Secret) *apicorev1.Secret {
+	gwSecret := &apicorev1.Secret{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: apicorev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name:      GatewaySecretName,
+			Namespace: istioNamespace,
+		},
+		Data: map[string][]byte{
+			"tls.crt": rootSecret.Data["tls.crt"],
+			"tls.key": rootSecret.Data["tls.key"],
+			"ca.crt":  rootSecret.Data["ca.crt"],
+		},
+	}
+	return gwSecret
+}
+
+func WatchChangesOnRootCertificate(clientset *kubernetes.Clientset, gatewaySecretHandler *GatewaySecretHandler,
+	log logr.Logger,
+) {
+	secretWatch, err := clientset.CoreV1().Secrets(istioNamespace).Watch(context.Background(), apimetav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(apimetav1.ObjectNameField, kcpRootSecretName).String(),
+	})
+	if err != nil {
+		log.Error(err, "unable to start watching root certificate")
+		panic(err)
+	}
+
+	for event := range secretWatch.ResultChan() {
+		rootCASecret, _ := event.Object.(*apicorev1.Secret)
+
+		switch event.Type {
+		case watch.Added:
+			fallthrough
+		case watch.Modified:
+			err := gatewaySecretHandler.ManageGatewaySecret(rootCASecret)
+			if err != nil {
+				log.Error(err, "unable to manage istio gateway secret")
+			}
+		case watch.Deleted:
```

Comment:
if it's deleted, we don't clean up the secret?

Reply:
thats is an undefined state. it should never be deleted AFAIK. Even if there is a workflow in the cefrt manager to delete it, it would be added again, that's when we'll tigger the update again. Note that secret being deleted doesn't mean that the copied gateway secret is invalid. It expires only at the `notAfter` date

Reply:
then please write a code comment here to explain this case.